pallets/delegation/src/lib.rs

```diff
@@ -334,58 +352,69 @@ impl<T: Trait> Module<T> {
 	}
 
 	/// Revoke a delegation and all of its children recursively
-	/// FIXME: Add bound
 	fn revoke(
 		delegation: &T::DelegationNodeId,
 		sender: &T::AccountId,
-		max_revocations: u64,
-	) -> Result<u64, DispatchError> {
-		let mut revocations: u64 = 0;
+		max_revocations: u32,
+	) -> Result<(u32, Weight), DispatchError> {
+		let mut revocations: u32 = 0;
+		let mut consumed_weight: Weight = 0;
 		// retrieve delegation node from storage
 		let mut delegation_node =
 			<Delegations<T>>::get(*delegation).ok_or(Error::<T>::DelegationNotFound)?;
+		consumed_weight += T::DbWeight::get().reads(1);
 
 		// check if already revoked
 		if !delegation_node.revoked {
 			// first revoke all children recursively
-			revocations += Self::revoke_children(delegation, sender, max_revocations)?;
+			Self::revoke_children(delegation, sender, max_revocations - 1).map(|(r, w)| {
+				revocations += r;
+				consumed_weight += w;
+			})?;
 
 			// if we run out of revocation gas, we only revoke children. The tree will be changed but is still valid.
 			if revocations < max_revocations {
 				// set revoked flag and store delegation node
 				delegation_node.revoked = true;
 				<Delegations<T>>::insert(*delegation, delegation_node);
+				consumed_weight += T::DbWeight::get().writes(1);
 				// deposit event that the delegation has been revoked
 				Self::deposit_event(RawEvent::DelegationRevoked(sender.clone(), *delegation));
 				revocations += 1;
 			} else {
 				return Err(Error::<T>::ExceededRevocationBounds.into());
 			}
 		}
-		Ok(revocations)
+		Ok((revocations, consumed_weight))
 	}
 
 	/// Revoke all children of a delegation
 	fn revoke_children(
 		delegation: &T::DelegationNodeId,
 		sender: &T::AccountId,
-		max_revocations: u64,
-	) -> Result<u64, DispatchError> {
-		let mut revocations: u64 = 0;
+		max_revocations: u32,
+	) -> Result<(u32, Weight), DispatchError> {
+		let mut revocations: u32 = 0;
+		let mut consumed_weight: Weight = 0;
 		// check if there's a child vector in the storage
 		if <Children<T>>::contains_key(delegation) {
 			// iterate child vector and revoke all nodes
 			let children = <Children<T>>::get(delegation);
+			consumed_weight += T::DbWeight::get().reads(1);
+
 			for child in children {
 				let remaining_revocations = max_revocations.saturating_sub(revocations);
 				if remaining_revocations > 0 {
-					revocations += Self::revoke(&child, sender, remaining_revocations)?;
+					Self::revoke(&child, sender, remaining_revocations).map(|(r, w)| {
+						revocations += r;
+						consumed_weight += w;
+					})?;
```

Comment:
wouldn't that work too? 😁 But the map is also a very cool solution.

```suggestion
					let (r, w) = Self::revoke(&child, sender, remaining_revocations)?;
					revocations += r;
					consumed_weight += w;
```

Reply:
That would work too. I would like to keep the map though, no need to instantiate two new vars right?